Lib/inspect.py

```diff
@@ -2151,7 +2279,7 @@ def _signature_from_builtin(cls, func, skip_bound_arg=True):
 
 
 def _signature_from_function(cls, func, skip_bound_arg=True,
-                             globalns=None, localns=None):
+                             globals=None, locals=None, eval_str=ONLY_IF_STRINGIZED):
```

Comment:
This may have backward compatibility problems: `inspect.signature` will start raising on 3.10 for a function that has all string annotations but uses names that aren't in the provided namespace. I feel like using `eval_str=False` here is safer.

Reply:
Before PEP 563 was reverted, inspect.signature() called typing.get_type_hints() to get the annotations for the function, which also called eval() on string annotations and could raise exceptions.  I was emulating this behavior.

Also, this is why the eval_str parameter exists.  Yes, it may start raising an exception for users in this situation, but it also allows for an easy fix.

There are good arguments to be made here for a default of False, and also for a default of ONLY_IF_STRINGIZED.  A default of ONLY_IF_STRINGIZED is bad for users who

- turn on stringized annotations,
- use annotations of values that aren't defined in their namespace, *and*
- examine the signature at runtime.

But it's better for literally everyone else, which is the vast majority of users.  So, as much as I don't like introducing a breaking change, I still think a default of ONLY_IF_STRINGIZED is the better choice.

Reply:
> Before PEP 563 was reverted, inspect.signature() called typing.get_type_hints() to get the annotations for the function, which also called eval() on string annotations and could raise exceptions. I was emulating this behavior.

More specifically, the implementation was like this;

```py
def _get_type_hints(func):
    try:
        return typing.get_type_hints(func)
    except Exception:
        # First, try to use the get_type_hints to resolve
        # annotations. But for keeping the behavior intact
        # if there was a problem with that (like the namespace
        # can't resolve some annotation) continue to use
        # string annotations
        return func.__annotations__
```

Reply:
Huh.  I specifically experimenting with typing.get_type_hints() and getting it to raise a NameError.  Is that catchall except statement recent?

Reply:
> Huh. I specifically experimenting with typing.get_type_hints() and getting it to raise a NameError. Is that catchall except statement recent?

No, it was there since the PEP 563 became default.

Reply:
Oh, okay.  So, what I'm emulating is the behavior of typing.get_type_hints() as it shipped in 3.9 (and presumably earlier versions).

    from __future__ import annotations

    def foo(a:int, b:flarg): pass

    import typing
    print(typing.get_type_hints(foo))

That raises a `NameError` in Python 3.9.4.

Reply:
FWIW, ``typing.get_type_hints`` also raises a ``NameError`` for your code example in the latest 3.10 from ``master``.

Reply:
Good.  I think that's the right choice for `typing.get_type_hints()`, and for `inspect.get_annotations()` and `inspect.signature()` too.

Note that, unlike `typing.get_type_hints()`, `inspect.get_annotations()` and `inspect.signature()` give the caller control over whether or not `eval()` is called.  If your annotations are stringized, and you only want to see the strings anyway, you can call them like so:

    ann = inspect.get_annotations(o, eval_str=False)
    sig = inspect.get_signature(o, eval_str=False)

Reply:
I've changed the default to eval_str=False.